### ISSUE TITLE
How to access scoped services from middleware

### ISSUE THREAD

**Issue by Manthazar** [Azure/azure-functions-dotnet-worker#836]
Hi all,

I am attempting to implement an authentication middleware. Once the auth is completed, I need to store the user context, such as the token in a scoped service (for later potential usage).

The middleware is a singleton per process, so I shall not resolve scoped services in its constructor (because the scope would then be falsely global).

Hence I attempted to use the FunctionContext to resolve the scoped service:
```
public class MyMiddleware : ... 
{
  public Task Invoke (FunctionContext context, ...)
  {
    context.InstanceServices.GetRequiredService<IUserContext>(); // Builds an instance of the user context for each request
  }
}
```
This correctly creates an instance for each request. However, a later attempt to inject the context through a constructor creates **ANOTHER** instance of it:
```
public class ActiveDirectoryLookup : IActiverDirectoryLookup
{
  public ActiveDirectoryLookup (IUserContext userContext) {... } // Results in **another** instance for the same request
}
```

- Both services are registered with scope, i.e. 
```
  services.AddScoped<IUserContext, UserContext>();
  services.AddScoped<IAzureDirectoryService, AzureActiveDirectoryService>();

```
- The scope however seems to be similar, but different (both are by request, but not equal)

**Two questions therefore:**
- How to access scoped services from middleware (if not through function context)
- How to pass data from middleware to a (business service), if a scoped service is not the way?

Aron

**Comment by Manthazar** [comment]
This is more a note (as it doesn't seem to relate to my issue, but the following in DefaultFunctionContext is wrong:
 ```
public override IServiceProvider InstanceServices
        {
            get
            {
                if (_instanceServicesScope == null)
                {
                    _instanceServicesScope = _serviceScopeFactory.CreateScope();
                    _instanceServices = _instanceServicesScope.ServiceProvider;
                }

                return _instanceServices!;
            }

            set => _instanceServices = value;
        }
```

If the public setter is used before the getter was used, the assigned value will be ignored and replaced on the get (since internalServicesScope is still null then)

**Comment by Manthazar** [comment]
I thinks my issue is not with middleware/ azure functions. The fishy is swimming in dependency injection space. 
(I tried to reproduce the case within the middleware example and it was fine!)

**Comment by fabiocav** [comment]
@Manthazar I'm assuming you've looked at the sample as a reference and that was working as expected, correct?

From your sample above, can you please clarify:
- How you're registering `IActiverDirectoryLookup`
- How you're consuming `IActiverDirectoryLookup`

If you can have a small repro in a GitHub repo and share that, that would be the best option as it would enable us to take a closer look at your specific scenario.

Thanks!

**Comment by Manthazar** [comment]
Yes, I tried to reproduce it in the example and failed. That has led me to the conclusion that the fish is not in Azure Function Worker water. 

After that, I had checked the lifetime of that service and it was _scoped_. Went down the rabbit hole and found that the caller was also _scoped_, but it has received a service provider supposedly being a _singleton_ (or in different scope) ultimately  turning the cache instance to a singleton. 

Anyway, I changed the orchestration in the following to not mess with timelines:
```
BEFORE: roleProvider (scoped) --> directory-cache (singleton) --> Azure Directory Service (scoped) 
AFTER:  roleProvider (scoped) --> directory-lookup mediator (scoped) 
                                                    --> On valid cache entries: directory-cache (singleton)
                                                    --> On cache expiration: Azure Directory Service (scoped) 
```
Bottom line is that singletons should **never be in the middle** of a call chain -only at the end.

It is a bit surprising though that the code works with Azure Function NetCore 3.1 though -there must be some other dark magic* going on. 

Sorry for wasting your time!

*The Azure Directory lookup uses there the IHttpContextAccess interface which is no longer available in the new version. That magic is dark because one would like to thing that it should fail for the same/ similar reason, but it doesn't.

**Comment by fabiocav** [comment]
No worries! Glad you found the root cause.

One thing that might help identify similar issues in the future is the enablement of scope validation. 

Something similar to what is described [here](https://github.com/Azure/azure-functions-dotnet-worker/issues/811#issuecomment-1051052233)
